common/http/platform/beast/http.cpp

```diff
@@ -321,6 +329,83 @@ error::Error Client::AsyncCall(
 	return error::NoError;
 }
 
+error::Error Client::HandleProxySetup() {
+	secondary_req_.reset();
+
+	if (request_->address_.protocol == "http") {
+		socket_mode_ = SocketMode::Plain;
+
+		if (http_proxy_ != "" && !HostNameMatchesNoProxy(request_->address_.host, http_proxy_)) {
+			// Make a modified proxy request.
+			BrokenDownUrl proxy_address;
+			auto err = BreakDownUrl(http_proxy_, proxy_address);
+			if (err != error::NoError) {
+				return err.WithContext("HTTP proxy URL is invalid");
+			}
+			if (proxy_address.path != "" && proxy_address.path != "/") {
+				return MakeError(
+					InvalidUrlError, "A URL with a path is not legal for a proxy address");
+			}
+
+			request_->address_.path = request_->address_.protocol + "://" + request_->address_.host
+									  + ":" + to_string(request_->address_.port)
+									  + request_->address_.path;
+			request_->address_.host = proxy_address.host;
+			request_->address_.port = proxy_address.port;
+			request_->address_.protocol = proxy_address.protocol;
+
+			if (proxy_address.protocol == "https") {
+				socket_mode_ = SocketMode::Tls;
+			} else if (proxy_address.protocol == "http") {
+				socket_mode_ = SocketMode::Plain;
+			} else {
+				// Should never get here.
+				assert(false);
+			}
+		}
+	} else if (request_->address_.protocol == "https") {
+		socket_mode_ = SocketMode::Tls;
+
+		if (https_proxy_ != "" && !HostNameMatchesNoProxy(request_->address_.host, https_proxy_)) {
+			// Save the original request for later, so that we can make a new request
+			// over the channel established by CONNECT.
+			secondary_req_ = std::move(request_);
+
+			request_ = make_shared<OutgoingRequest>();
+			request_->SetMethod(Method::CONNECT);
+			BrokenDownUrl proxy_address;
+			auto err = BreakDownUrl(https_proxy_, proxy_address);
+			if (err != error::NoError) {
+				return err.WithContext("HTTPS proxy URL is invalid");
+			}
+			if (proxy_address.path != "" && proxy_address.path != "/") {
+				return MakeError(
+					InvalidUrlError, "A URL with a path is not legal for a proxy address");
+			}
```

Comment:
shared above...? Am I missing something?

Reply:
I mean that this 9 lines are the exact same as above in https://github.com/mendersoftware/mender/pull/1433/files#diff-7100548573dbc4eefb8ff288d50eb33b3c038a36e5573de55fa670b1f0221c46R380-R388 -> we could refactor in a tiny helper function

Reply:
It's not that easy. Some parts are the same, but there are subtle differences sprinkled all over it: Several references to "https" instead of "http", the lower path switches out the `request_` member, and the path construction is different. It wouldn't be impossible to break some of it out, but all ways that I've found significantly hurt readability.

Reply:
You are right :see_no_evil: